implement epitope similarity regularization

timcyu opened the pull request:
This PR implements the epitope similarity regularization idea in #124. It is not ready to merge yet. 

I've included a test for epitope similarity regularization in `test_gradients.py`. Things match up well for the most part, but I noticed the difference between the numerical and analytical gradient is sometimes greater than the threshold allowed. I changed the difference allowed to `1e-6` (previously `1e-7`) but unsure if this is appropriate. The gradients differ by as much as `2.4e-07` (numerical gradient: `2.4e-07`, analytical gradient: 0.0).

I've also included a notebook showing the regularization helps deconvolve Frances' H3 HA mAb cocktail DMS data. This is a cocktail containing two mAbs that bind two distinct epitopes on HA. When you fit an `n_epitopes=2` model with default parameters (no epitope similarity regularization), you get two identical escape profiles. 
<img width="819" alt="without reg" src="https://user-images.githubusercontent.com/36126324/194223343-0977d802-0089-49ef-b4ff-054b42efab8d.png">

With the regularization, it's a different story! The "polyclonal serum" is successfully deconvolved.
<img width="808" alt="with reg" src="https://user-images.githubusercontent.com/36126324/194223664-4ea8f8dd-0fea-46d5-85ee-ca221d04b59a.png">

Additionally, there's a notebook (which shouldn't be merged) that fits the same model in `fit_RBD.ipynb` but with a little bit of epitope similarity regularization. Interestingly, this ruins the result by making all the escape values for epitope 1 ~ 0. My guess is this is happening because many of the epitope 1 escape sites have overlapping escape values at other epitopes. This suggests to me that the default penalty should be `0` unless absolutely needed (like in above cocktail dataset).

Lastly, the equations for the penalty and its gradient are in `optimization.ipynb`. I think the notation could probably be improved to make things more accurate and clear.

jbloom reviewed (changes requested):
@timcyu, good start. Here are some requests just based on the equations. Check these are right, fix the description, and harmonize the code with them. Then let me know and I'll look more.

- [ ] in `optimization.ipynb` in the formula for $R_{similarity}$, the $S_k$ values should not depend on $i$ or $j$ since the mutations at a site don't depend on epitope. So replace $S_{k,i}$ and $S_{k,j}$ with $S_k$ 
- [ ] in `optimization.ipynb` in the formula for the derivative of $R_{similarity}$, make the same change listed above: $S_k$ does not depend on $i$ or $j$. Additionally, you don't need the sum of $i$ and $j$ in the derivatives as all the values not equal to $e$ are zero, so just drop $i$ and $j$ altogether and replace them with $e$. Likewise I'm not sure that derivative should have the sum over $k$, as we only care about the site that $m$ belongs to. So use some notation that indicates that.

jbloom commented:
@timcyu, looks good. My only comment is in the derivative, move the sum over $j \ne e$ to the right of $\beta_{m,e}$ since the $\beta_{m,e}$ term doesn't depend on $j$.

timcyu commented:
@jbloom it's ready for your review again!

jbloom reviewed (changes requested):
@timcyu, another change to `optimization.ipynb`.

- [x] The text is out of date in terms of the textual descriptions. At the top of the notebook, it mentions two regularizations. Then in the *Regularization* section it mentions three (but not the new one you added). I'd say get rid of the description of regularization at the top of the page and just refer to the *Regularization* section, but then in the introductory text to that section make sure it's up-to-date in describing all the implemented regularizations.

jbloom reviewed (changes requested):
@timcyu, a few comments on the implementation in `polyclonal.py`. For the most part, looks really good. But I'd suggest the following:

 - [ ] in `_reg_similarity` do not raise an error if there is only one epitope, instead just return the weight as zero. The reason being that in the future we may want to make the default weight here >0, and if so we don't want an error anytime there is more than one epitope, or to need two different defaults. I think it's sufficient for the docs for `reg_similiarity_weight` to just say "has no effect when there is only one epitope."
 - [ ] the code can be more efficient. The more you can do things once using `numpy` arrays and reduce calculations in loops, the better. For instance, right now you re-calculate `norm_ei` and `norm_ej` repeatedly when there are >2 combinations. Think judiciously about where your summation is. For instance, in your loops, I think everything depends on the site but only some things depend on the epitope. So the site index should be first and don't recompute things in inner loops.
 - [ ] I honestly don't understand what all the `zero_idx` stuff is doing, but it looks super complicated for implementing what should be a simple formula. See if you can simplify. Related to that, `list(set(range(len(self.epitopes)))` looks wrong to me. Sets are unordered, so why you would turn a list into a set and then back into a list is unclear to me, but it's not going to give a reproducible results since sets are not ordered.

Try simplifying this code and then I can take a more careful look.

timcyu commented:
@jbloom addressed your comments. I decided to re-write the code using some of the matrix tricks in Will's colab notebook so it is far more efficient now. Let me know what you think.

jbloom commented:
@timcyu, I made a few small tweaks in commit 51fcdb2 (see full commit message for details). Otherwise this looks good, but let's implement other approach (absolute value approximation: https://github.com/jbloomlab/polyclonal/issues/124#issuecomment-1272055558) and see which we prefer before deciding which pull request to merge.

timcyu commented:
@jbloom I would suggest merging this regularization instead of the absolute value approximation. Reason is that both strategies can successfully deconvolve the mAb cocktail dataset, but the absolute value approximation is **much** more sensitive to the regularization weight, whereas I could reproduce the same result (shown above) with a larger range of weights with this squared euclidean penalty. 

We discussed this last week and I've included the relevant slides [here](https://docs.google.com/presentation/d/1bmfzye_EFG_Ar5IWuclS2oorsb1o3niTaDwzr3cAqBw/edit?usp=sharing) for reference. Let me know if you have any final thoughts!

jbloom commented:
@timcyu, can you experiment on what happens if you drop `reg_activity_weight` a lot, maybe even pushing it to 0? Particularly on the simulated RBD dataset. A lot of the issue seems to be that the typical non-epitope site is being fit to something more like ~0.5 than 0, which may be why you lose the third epitope. I wonder if that is driving by the regularization on activity which is moving some of that value into the escape values.

timcyu commented:
@jbloom dropping `reg_activity_weight` has no effect on both the simulated RBD and mAb cocktail datasets.

jbloom reviewed (changes requested):
@timcyu, great, can you just do two more thing:
 - update the version in `__init__.py` to `2.dev5`
 - update the CHANGELOG to put an entry under `2.dev5` to describe what you've added.
